### PR TITLE
Added hints & fixups chapters 2.1-2.4 and 3.1.

### PR DESCRIPTION
**Pull request by jachiang**

### CODE REVIEW

File: 2.1-Segwit-Version-1.ipynb [hunk 1]
```diff
@@ -552,6 +558,13 @@
    "source": [
     "test.shutdown()"
    ]
+  },
```
Comment (jnewbery) [inline 2.1-Segwit-Version-1.ipynb]: Why are you adding this cell?

File: 2.1-Segwit-Version-1.ipynb [hunk 2]
```diff
@@ -411,17 +410,20 @@
...
+    "privkey2_c = # TODO: Implement.\n",
+    "pubkey1_c = # TODO: Implement.\n",
+    "pubkey2_c = # TODO: Implement.\n",
     "\n",
     "# Create a segwit v1 address for the MuSig aggregate pubkey\n",
-    "# Use segwit_addr.encode(\"bcrt\", witness_version, witness_program)\n",
-    "program_musig =  # TODO: implement\n",
-    "address_musig =  # TODO: implement\n",
+    "# Tip: address = program_to_witness(version_int, program_bytes)\n",
+    "program_musig = # TODO: Implement.\n",
```
Comment (jnewbery) [inline 2.1-Segwit-Version-1.ipynb]: why this change?

File: 2.2-TapTweak.ipynb [hunk 3]
```diff
@@ -15,7 +15,7 @@
     "from test_framework.key import ECKey, ECPubKey, SECP256K1_ORDER, generate_schnorr_nonce\n",
     "from test_framework.messages import CTransaction, COutPoint, CTxIn, CTxOut, CScriptWitness, CTxInWitness, ser_string\n",
     "from test_framework.musig import generate_musig_key, aggregate_schnorr_nonces, sign_musig, aggregate_musig_signatures\n",
-    "from test_framework.script import  CScript, OP_1, OP_CHECKSIG, TaprootSignatureHash, TapLeaf, TapTree, Node"
+    "from test_framework.script import  CScript, OP_1, OP_CHECKSIG, TaprootSignatureHash, TapLeaf, TapTree, Node\n"
```
Comment (jnewbery) [inline 2.2-TapTweak.ipynb]: why this new newline?

File: 2.2-TapTweak.ipynb [hunk 4]
```diff
@@ -122,24 +122,24 @@
     "privkey1_c = privkey1.mul(c_map[pubkey1])\n",
     "pubkey0_c = pubkey0.mul(c_map[pubkey0])\n",
     "pubkey1_c = pubkey1.mul(c_map[pubkey1])\n",
-    "    \n",
+    "\n",
     "# Tweak musig public key\n",
+    "# Tip: ECPubKey.tweak_add()\n",
     "tweak = random.randrange(1, SECP256K1_ORDER)\n",
-    "agg_pubkey_tweaked =  # TODO: implement\n",
+    "agg_pubkey_tweaked = # TODO: implement\n",
```
Comment (jnewbery) [inline 2.2-TapTweak.ipynb]: why this change?

File: 2.2-TapTweak.ipynb [hunk 5]
```diff
@@ -122,24 +122,24 @@
...
     "\n",
     "# Signing and signature aggregation\n",
     "msg = hashlib.sha256(b'msg').digest()\n",
     "\n",
     "# Sign individually and then aggregate signatures\n",
-    "# method: use sign_musig() and aggregate_musig_signatures()\n",
-    "sig_agg =  # TODO: implement\n",
+    "# Tip: sign_musig(private_key, nonce_key, nonce_point, public_key, msg)\n",
+    "# Tip: aggregate_musig_signatures(signature_list)\n",
+    "sig_agg = # TODO: implement\n",
```
Comment (jnewbery) [inline 2.2-TapTweak.ipynb]: why remove the space here?

File: 2.2-TapTweak.ipynb [hunk 6]
```diff
@@ -302,20 +302,20 @@
    "outputs": [],
    "source": [
     "# Example key pair\n",
-    "privkey = ECKey().set(102118636618570133408735518698955378316807974995033705330357303547139065928052, True)\n",
+    "privatekey = ECKey().set(102118636618570133408735518698955378316807974995033705330357303547139065928052, True)\n",
```
Comment (jnewbery) [inline 2.2-TapTweak.ipynb]: Can we please just standardise on pubkey/privkey rather than have different variable names everywhere?

File: 2.2-TapTweak.ipynb [hunk 7]
```diff
@@ -302,20 +302,20 @@
...
     "\n",
     "# Example tweak\n",
     "taptweak = bytes.fromhex('2a2fb476ec9962f262ff358800db0e7364287340db73e5e48db36d1c9f374e30')\n",
     "\n",
     "# Tweak the public key\n",
-    "# Method: use tweak_add()\n",
-    "taproot_pubkey =  # TODO: implement\n",
-    "taproot_pubkey_b =  # TODO: implement\n",
+    "# Tip: use tweak_add()\n",
+    "taproot_pubkey = # TODO: implement\n",
```
Comment (jnewbery) [inline 2.2-TapTweak.ipynb]: why are you changing all of these?

File: 2.2-TapTweak.ipynb [hunk 8]
```diff
@@ -302,20 +302,20 @@
...
     "internal_pubkey = privatekey.get_pubkey()\n",
     "\n",
     "# Example tweak\n",
     "taptweak = bytes.fromhex('2a2fb476ec9962f262ff358800db0e7364287340db73e5e48db36d1c9f374e30')\n",
     "\n",
     "# Tweak the public key\n",
-    "# Method: use tweak_add()\n",
-    "taproot_pubkey =  # TODO: implement\n",
-    "taproot_pubkey_b =  # TODO: implement\n",
+    "# Tip: use tweak_add()\n",
```
Comment (jnewbery) [inline 2.2-TapTweak.ipynb]: I think "Method:" is fine.

File: 2.2-TapTweak.ipynb [hunk 9]
```diff
@@ -429,6 +433,13 @@
     "# Shutdown\n",
     "test.shutdown()"
    ]
+  },
+  {
```
Comment (jnewbery) [inline 2.2-TapTweak.ipynb]: why?

**Comment by jachiang** [comment]
Updated and rebased on master.

File: 2.1-Segwit-Version-1.ipynb [hunk 10]
```diff
@@ -411,17 +410,20 @@
     "pubkey1 = privkey1.get_pubkey()\n",
     "pubkey2 = privkey2.get_pubkey()\n",
     "\n",
-    "# Generate MuSig key\n",
-    "# Method: use generate_musig_key()\n",
-    "\n",
+    "# Generate a 2-of-2 aggregate MuSig key\n",
+    "# Method: generate_musig_key([ECPubKey_list])\n",
```
Comment (elichai) [inline 2.1-Segwit-Version-1.ipynb]: nit, I would drop the square parenthesis, I think the variable name tells enough.

File: solutions/2.1-Segwit-Version-1-Solutions.txt [hunk 11]
```diff
@@ -7,7 +7,7 @@ pubkey1 = privkey1.get_pubkey()
 pubkey2 = privkey2.get_pubkey()
 
 # Generate a 2-of-2 aggregate MuSig key
-# Method: use generate_musig_key()
+# Method: generate_musig_key([ECPubKey_list])
```
Comment (elichai) [inline solutions/2.1-Segwit-Version-1-Solutions.txt]: same nit

File: solutions/2.1-Segwit-Version-1-Solutions.txt [hunk 12]
```diff
@@ -17,11 +17,11 @@ pubkey1_c = pubkey1 * c_map[pubkey1]
 pubkey2_c = pubkey2 * c_map[pubkey2]
 
 # Create a segwit v1 address for the MuSig aggregate pubkey
-# Use segwit_addr.encode("bcrt", witness_version, witness_program)
+# Method: address = program_to_witness(version_int, program_bytes)
```
Comment (elichai) [inline solutions/2.1-Segwit-Version-1-Solutions.txt]: Better :) Thanks!

File: 2.1-Segwit-Version-1.ipynb (hunk 10 above)
Comment (jachiang) [reply 2.1-Segwit-Version-1.ipynb]: Thx. Fixed

File: 2.1-Segwit-Version-1.ipynb [hunk 13]
```diff
@@ -411,17 +410,20 @@
...
+    "privkey2_c =  # TODO: Implement\n",
+    "pubkey1_c =  # TODO: Implement\n",
+    "pubkey2_c =  # TODO: Implement\n",
     "\n",
     "# Create a segwit v1 address for the MuSig aggregate pubkey\n",
-    "# Use segwit_addr.encode(\"bcrt\", witness_version, witness_program)\n",
-    "program_musig =  # TODO: implement\n",
-    "address_musig =  # TODO: implement\n",
+    "# Method: address = program_to_witness(version_int, program_bytes)\n",
+    "program_musig =  # TODO: Implement\n",
```
Comment (jnewbery) [inline 2.1-Segwit-Version-1.ipynb]: Please remove all these capitalization changes from this PR

**Comment by jnewbery** [comment]
I've rebased on master and force-pushed a bunch of minor changes
